### DISCUSSION TITLE
docs: clarify internal em.getContext() usage

### DISCUSSION THREAD

**Discussion by jahudka** [mikro-orm/mikro-orm#3122] [General]
The docs state in multiple places that the `RequestContext` helper is used transparently under the hood, e.g.:

> MikroORM will always use request specific (forked) entity manager if available [...]

> This method is used under the hood automatically, so we should not need it.

> Internally all EntityManager methods that work with the Identity Map (e.g. em.find() or em.getReference()) first call em.getContext() to access the contextual fork.

This is typically achieved using the `EntityManager.getContext()` method, which also takes care of obtaining a transaction-specific entity manager using a mechanism similar to the `RequestContext` helper. But in fact this doesn't work as transparently as advertised, which is especially evident when one doesn't use repositories to access entities:

- The base `EntityRepository` provides to its descendants (and internally uses) a _getter_ for the entity manager, which calls `getContext()` to obtain the correct EM instance, so when repositories are used, we're somewhat shielded from this.
- `EntityManager.find()` uses helper methods to access the UoW and the entity factory; these methods internally call `getContext()`. But other places in the method work with the current EM instance directly, which leads to hard-to-find bugs - for example the query is run using the `transactionContext` property of the current EM instance without resolving the transaction-scoped EM first, so unless one is working with a transaction-scoped EM instance in the first place, `find()` will run _outside_ the transaction. I'm not sure about other things the method does, such as the `tryCache()` call.
- `EntityManager.findOne()` checks the lock mode against the current EM instance; if this isn't the transaction-scoped instance and a lock is requested, the call will fail with an error stating that an open transaction is required for locks.

I'm unsure whether to report this as a bug or a documentation request because the code is ambivalent about whether the contextual EM will be used or not (sometimes it is, sometimes it isn't), so perhaps this is expected and intended and people should just use repositories for everything (which I normally would as well, but I'm porting a legacy project from TypeORM and I have custom "manager" classes which I haven't needed to convert to repositories yet). But in any case this leads to _extremely_ hard to debug problems where queries are run at the right time but on the wrong connection and errors are being thrown about an open transaction being required when literally two lines in the log before this you see the `begin` query being issued.. so I'd say that it should at least be more thoroughly documented.

**Comment by B4nan** [comment]
The method should have been marked as internal, its not intended for users at all, basically only for debugging purposes of context issues. So I am not really interested in deep diving to documenting such things (there are many much more important things I'd like to document, this is purely implementation detail). I believe I already explained the generals in the identity map docs section, users should not care more about it than that.

The contextual EM should be used always where needed - some parts of the ORM are context aware (UoW, EF), some are not (driver, connection, platform, metadata) - there is no sense in working with contextual fork if all we need from it is to access shared singleton instances. 

If you see bugs, please provide reproducitons so we can iron them out. Everything needs to work the same way with EM as with repositories.

**Comment by B4nan** [comment]
I believe [this commit](https://github.com/mikro-orm/mikro-orm/commit/cc6d59b9a96fbfd4106e37c21814ded84b56588a) should help with the issues you were describing, it basically ensures same behaviour as if you would use `em.getContext()` in from of all public EM methods (so same as if the getter from repository class would be used). I also realized there was a bug in `getContext()` method as it always preferred the transactional context even for EMs with `useContext = false` (so basically forks). 

Related discussion with a repro for one case is in #3271
